spanner/transaction.go

```diff
@@ -1071,21 +1127,94 @@ func (t *ReadWriteTransaction) batchUpdateWithOptions(ctx context.Context, stmts
 }
 
 // acquire implements txReadEnv.acquire.
+// This will make sure that only one operation will be running with TransactionSelector::begin option
+// in a ReadWriteTransaction by changing the state to init, all other operations will wait for state
+// to become active/closed. If state is active transactionID is already set, if closed returns error.
 func (t *ReadWriteTransaction) acquire(ctx context.Context) (*sessionHandle, *sppb.TransactionSelector, error) {
-	ts := &sppb.TransactionSelector{
-		Selector: &sppb.TransactionSelector_Id{
-			Id: t.tx,
-		},
+	for {
+		t.mu.Lock()
+		switch t.state {
+		case txClosed:
+			t.mu.Unlock()
+			return nil, nil, errTxClosed()
+		case txNew:
+			// State transit to txInit so that only one TransactionSelector::begin
+			// is accepted.
+			t.state = txInit
+			sh := t.sh
+			ts := &sppb.TransactionSelector{
+				Selector: &sppb.TransactionSelector_Begin{
+					Begin: &sppb.TransactionOptions{
+						Mode: &sppb.TransactionOptions_ReadWrite_{
+							ReadWrite: &sppb.TransactionOptions_ReadWrite{},
+						},
+					},
+				},
+			}
+			t.mu.Unlock()
+			return sh, ts, nil
+		case txInit:
+			if t.tx == nil {
+				// Wait for a transaction ID to become ready.
+				txReadyOrClosed := t.txReadyOrClosed
+				t.mu.Unlock()
+				select {
+				case <-txReadyOrClosed:
+					// Need to check transaction state again.
+					continue
+				case <-ctx.Done():
+					// The waiting for initialization is timeout, return error
+					// directly.
+					return nil, nil, errTxInitTimeout()
+				}
+			}
+			// If first statement with TransactionSelector::begin succeeded, t.state should have been changed to
+			// txActive, so we can just continue here.
+			continue
+		case txActive:
+			sh := t.sh
+			ts := &sppb.TransactionSelector{
+				Selector: &sppb.TransactionSelector_Id{
+					Id: t.tx,
+				},
+			}
+			t.mu.Unlock()
+			return sh, ts, nil
+		}
+		state := t.state
+		t.mu.Unlock()
+		return nil, nil, errUnexpectedTxState(state)
 	}
+}
+
+func (t *ReadWriteTransaction) getTransactionSelector() *sppb.TransactionSelector {
 	t.mu.Lock()
 	defer t.mu.Unlock()
-	switch t.state {
-	case txClosed:
-		return nil, nil, errTxClosed()
-	case txActive:
-		return t.sh, ts, nil
+	if t.state == txActive {
+		return &sppb.TransactionSelector{
+			Selector: &sppb.TransactionSelector_Id{
+				Id: t.tx,
+			},
+		}
+	}
+	return &sppb.TransactionSelector{
+		Selector: &sppb.TransactionSelector_Begin{
+			Begin: &sppb.TransactionOptions{
+				Mode: &sppb.TransactionOptions_ReadWrite_{
+					ReadWrite: &sppb.TransactionOptions_ReadWrite{},
+				},
+			},
+		},
 	}
-	return nil, nil, errUnexpectedTxState(t.state)
+}
+
+func (t *ReadWriteTransaction) setTransactionID(tx transactionID) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.tx = tx
+	t.state = txActive
+	close(t.txReadyOrClosed)
```

Comment:
This bit seems odd to me. Why is the channel being closed only to be reinitialized in the next line?

Reply:
Channel is closed to un-block other parallel routines waiting for transaction state to be Closed/Active, we are re-initialising is to new channel for the case of Closed so that when doing retry we don't access the closed channel from previous run.